notebook/terminal/terminalmanager.py

```diff
@@ -0,0 +1,156 @@
+"""A MultiTerminalManager for use in the notebook webserver
+- raises HTTPErrors
+- creates REST API models
+"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import warnings
+
+from datetime import timedelta
+from notebook._tz import utcnow, isoformat
+from terminado import NamedTermManager
+from tornado import web
+from tornado.ioloop import IOLoop, PeriodicCallback
+from traitlets import Integer, validate
+from traitlets.config import Configurable
+from ..prometheus.metrics import TERMINAL_CURRENTLY_RUNNING_TOTAL
+
+
+class TerminalManager(Configurable, NamedTermManager):
+    """  """
+
+    _culler_callback = None
+
+    _initialized_culler = False
+
+    cull_inactive_timeout = Integer(0, config=True,
+        help="""Timeout (in seconds) in which a terminal has been inactive and ready to be culled.
+        Values of 0 or lower disable culling."""
+                                    )
+
+    cull_interval_default = 300  # 5 minutes
+    cull_interval = Integer(cull_interval_default, config=True,
+        help="""The interval (in seconds) on which to check for terminals exceeding the inactive timeout value."""
+                            )
+
+    @validate('cull_interval')
+    def _cull_interval_validate(self, proposal):
+        value = proposal['value']
+        if value <= 0:
+            warnings.warn("Invalid value for 'cull_interval' detected ({}) - using default value ({}).".
+                          format(value, self.cull_interval_default))
```

Comment:
@kevin-bates, I saw your comment on logging here. `self.log.warn` doesn't work here? The class should be initialized already. I'm diving into traitlets now to see if I can get it to work.

Reply:
In the example below, I'm able to trigger logs from validate methods. This suggests that you should be able log within this validate method. Are you seeing an error? Or just no logs being emitted?

```python
import io
import logging
from traitlets.config import LoggingConfigurable
from traitlets import validate, Unicode


class A(LoggingConfigurable):

    a = Unicode()

    @validate('a')
    def _validate_a(self, proposal):
        # Fire a log warning in a validate method.
        self.log.warning("Warning seen!")
        return proposal



if __name__ == "__main__":
    # Setup logger and catch log output as io stream
    logger = logging.getLogger()
    out = io.StringIO()
    handler = logging.StreamHandler(out)
    logger.addHandler(handler)

    # Initialize the Logging Configurabel
    a = A(log=logger)

    # Setting the trait should fire the validate
    # method and trigger the warning to the log.
    a.a = "test"

    print(f"Log output: {out.getvalue()}")
```

```bash
>>> python traitlog.py                                                                                                                                   
Log output: Warning seen!
```

Reply:
I don't see anything logged.  Here's the output using `warnings.warn`...
```
[I 09:11:22.883 NotebookApp] It looks like you're running the notebook from source.
        If you're working on the Javascript of the notebook, try running
    
        npm run build:watch
    
        in another terminal window to have the system incrementally
        watch and build the notebook's JavaScript for you, as you make changes.
/Users/kbates/repos/oss/jupyter/notebook/notebook/terminal/terminalmanager.py:43: UserWarning: Invalid value for 'cull_interval' detected (-10) - using default value (300).
  format(value, self.cull_interval_default))
[I 09:11:23.255 NotebookApp] Serving notebooks from local directory: /Users/kbates/notebooks/alice
```
And here's that same sequence when switched to `self.log.warning`...
```
[I 09:13:39.518 NotebookApp] It looks like you're running the notebook from source.
        If you're working on the Javascript of the notebook, try running
    
        npm run build:watch
    
        in another terminal window to have the system incrementally
        watch and build the notebook's JavaScript for you, as you make changes.
[I 09:13:39.768 NotebookApp] Serving notebooks from local directory: /Users/kbates/notebooks/alice
```
One issue is that TerminalManager isn't derived from `LoggingConfigurable` - which I will update.  However, even after that change, still the same behavior (nothing logged).

Reply:
See my message below. I *think* this is grabbing the wrong logger object (python logging is always confusing to me).

Reply:
Addressed in later commits. We decided to keep the logging outside of traitlets validation due to odd behavior in class construction from terminado.